File: data_structures/ruby/lru_cache.rb
class LRUCacheInefficient 

    attr_reader :size, :capacity
    def initialize(capacity = 0)
        @capcity = 0
        @entries = {}
        @keys = []
        @size = @entries.size
    end

    def update_key(key)
        key_index = @keys.index(key)
        @keys[key_index] = nil
        @keys.push(key)
    end

    def set(key, value)
        if @entries.has_key?(key)
            @entries[key] = value 
            update_key(key)
        else
            if !remaining_capacity
                i = 0
                removed_key = @keys[i]
                while removed_key.nil?
                    i++
                    removed_key = @keys[i]
                end
                @entries.delete(key)
            end
            @entries[key] = value 
            @keys.push(key)
        end
    end

    def get(key)
        if @entries.has_key?(key)
            update_key(key)
            return @entries(key)
        else
            return nil
        end
    end

    def remaining_capacity
        capacity - size
    end
end

class LRUNode
    def initialize(key, value)
        @key = key
        @value = value
        @older = nil
        @newer = nil
    end
end

class LRUCacheEfficient < LRUCacheInefficient
    def initialize
        super();
        @head = nil
        @tail = nil
    end

    def update_key(key)
        existing_node = @entries[key]

        if existing_node.newer
            existing_node.newer.older = existing_node.older
        else
            @head = existing_node.older
        end

        if existing_node.older
            existing_node.older.newer = existing_node.newer
        else
            @tail = existing_node.newer
        end

        existing_node.older = @head 
        existing_node.newer = nil

        if @head
            @head.newer = existing_node
        end
        @head = existing_node
        if @tail.nil?
            @tail = existing_node
        end
    end

    def set(key, value)
        new_node = LRUNode.new(key, value)
        if @entries.has_key(key)
            @entries[key] = value
            update_key(key)
            return @entries[key]
        else 
            if !remaining_capacity
                deleted_key = @tail.key
                @tail = @tail.newer
                if @tail
                    @tail.older = nil
                end
                @entries.delete(deleted_key)
            end

            new_node.older = @head 
            if @head
                @head.newer = new_node
            end
            @head = new_node

            if @tail.nil?
                @tail = new_node
            end

            @entries[key] = new_node
        end
    end
end

File: data_structures/javascript/LRUCache.js
//Cache management algorithms are an entire field of Computer Science
//Removing an item from a cache: eviction
//Makes sense to add a callback for processing evicted items
//Cache size should be based on the expected available memory in your system


class LRUCacheInefficient {
    constructor(capacity = 0) {
        this.capacity = capacity;
        this.entries = {};
        this.keys = [];
        this.size = this.entries.size;
    }

    //O(this.keys.length)
    updateKey(key) {
        let keyIndex = this.keys.indexOf(key);
        this.keys[keyIndex] = undefined;
        this.keys.push(key);
    }

    //O(this.keys.length)
    set(key, value) {
        if (this.entries.hasOwnProperty(key)) {
            this.entries[key] = value;
            this.updateKey(key);
            return this.entries[key];
        } else {
            if (!this.remainingCapacity()) {
                let i = 0;
                let removedKey = this.keys[i];
                while (typeof removedKey === 'undefined') {
                    i++;
                    removedKey = this.keys[i];
                }
                delete this.entries[removedKey];
            }

            this.entries[key] = value;
            this.keys.push(key);
        }
    }

    //O(this.keys.length)
    get(key) {
        if (this.entries.hasOwnProperty(key)) {
            this.updateKey(key);
            return this.entries[key];
        } else {
            return null;
        }
    }

    //O(1)
    size() {
        return this.size;
    }

    //O(1)
    remainingCapacity() {
        return this.capacity - this.size;
    }
}

class LRUNode {
    constructor(key, value) {
        this.key = key;
        this.value = value;
        this.older = null;
        this.newer = null;
    }
}
class LRUCacheEfficient extends LRUCacheInefficient {
    constructor() {
        super();
        this.head = null;
        this.tail = null;
    }

    //O(1)
    updateKey(key) {
        const existingNode = this.entries[key];
        if (existingNode.newer) {
            existingNode.newer.older = existingNode.older;
        } else {
            this.head = existingNode.older;
        }

        if (existingNode.older) {
            existingNode.older.newer = existingNode.newer;
        } else {
            this.tail = existingNode.newer;
        }

        existingNode.older = this.head;
        existingNode.newer = null;
        if (this.head) {
            this.head.newer = existingNode;
        }
        this.head = existingNode;
        if (!this.tail) {
            this.tail = existingNode;
        }
    }

    //O(1)
    set(key, value) {
        const newNode = new LRUNode(key, value);
        if (this.entries.hasOwnProperty(key)) {
            this.entries[key] = value;
            this.updateKey(key);
            return this.entries[key];
        }

        if (!this.remainingCapacity()) {
            const deletedKey = this.tail.key;
            this.tail = this.tail.newer;
            if (this.tail) {
                this.tail.older = null;
            }
            delete this.entries[deletedKey];
        }

        newNode.older = this.head;
        if (this.head) {
            this.head.newer = newNode;
        }
        this.head = newNode;
        if (!this.tail) {
            this.tail = newNode;
        }
        this.entries[key] = newNode;
    }
}
